### PR TITLE
make io.pyi import everything from _io.py(i), like io.py does

### PR DESCRIPTION
**Pull request by matthiaskramm**

### CODE REVIEW

**Comment by matthiaskramm** [comment]
This PR got larger than expected. The basic motivation is to allow code like

```
io.BufferedReader(fp)
```

to pass type-checking. (Which it didn't before, since `io` didn't have a `BufferedReader`)

**Comment by gvanrossum** [comment]
Do you have code that directly imports `_io`? If not, maybe we should get rid of it completely, like we did in the PY3 stubs.

Anyway, I tested this without our internal code and it found one issue, io.BufferedIOBase is supposed to define `__enter__` and `__exit__`.

I am also still worried about some classes not inheriting from BinaryIO where they should. BinaryIO is abstract, we must ensure the abstractness or concreteness of various classes matches those in the stdlib.

**Comment by matthiaskramm** [comment]
> Do you have code that directly imports _io? 

We do, unfortunately. Some in internal code, but I'm also seeing occurrences in third party packages like pymock and astroid.

> I am also still worried about some classes not inheriting from BinaryIO where they should. BinaryIO is abstract, we must ensure the abstractness or concreteness of various classes matches those in the stdlib.

Right. Made `_IOBase` inherit from `BinaryIO`. (Note that `_TextIOBase` inherits from `_IOBase`, too, but is a `TextIO`. So you have a class that inherits both from `BinaryIO` and `TextIO`, which is a bit awkward. But that's also how it was modelled before this PR.)

**Comment by gvanrossum** [comment]
Seeing all the `# type: ignore` lines I wonder if it wouldn't be better to ignore the runtime/implementation inheritance between `_TextIOBase` and `_IOBase` and just claim (to the type system) that `_TextIOBase` doesn't inherit from `_IOBase`? Or perhaps construct a new fictitious class/type that represents the common methods like `seek()`?

**Comment by matthiaskramm** [comment]
I don't see why not. Let me look into it.

File: stdlib/2/_io.pyi [hunk 1]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
-    def readline(self, limit: int = ...) -> bytes: ...
-    def readlines(self, hint: int = ...) -> list[bytes]: ...
     def seek(self, offset: int, whence: int = ...) -> int: ...
     def seekable(self) -> bool: ...
     def tell(self) -> int: ...
     def truncate(self, size: Optional[int] = ...) -> int: ...
     def writable(self) -> bool: ...
-    def write(self, s: bytes) -> int: ...
+    def __exit__(self, type, value, traceback) -> bool: ...
+    # This just returns self:
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: Then why not use a self type?

File: stdlib/2/_io.pyi [hunk 2]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
     def readable(self) -> bool: ...
-    def readline(self, limit: int = ...) -> bytes: ...
-    def readlines(self, hint: int = ...) -> list[bytes]: ...
     def seek(self, offset: int, whence: int = ...) -> int: ...
     def seekable(self) -> bool: ...
     def tell(self) -> int: ...
     def truncate(self, size: Optional[int] = ...) -> int: ...
     def writable(self) -> bool: ...
-    def write(self, s: bytes) -> int: ...
+    def __exit__(self, type, value, traceback) -> bool: ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: Why did you remove the argument types here?

File: stdlib/2/_io.pyi [hunk 3]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
 class _BufferedIOBase(_IOBase):
-    def read1(self, n: int) -> str: ...
-    def read(self, n: int = ...) -> str: ...
+    def read1(self, size: int) -> bytes: ...
+    def read(self, size: int = ...) -> bytes: ...
     def readinto(self, buffer: bytearray) -> int: ...
-    def write(self, s: str) -> int: ...
+    def write(self, s: bytes) -> int: ...
     def detach(self) -> "_BufferedIOBase": ...
+    def __enter__(self) -> '_BufferedIOBase': ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: Looks odd to have double quotes on one line and single on the next. Since this is a stub, we can just remove all quotes anyway.

File: stdlib/2/_io.pyi [hunk 4]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
+    def __enter__(self) -> 'BytesIO': ...
 
 class _RawIOBase(_IOBase):
     def readall(self) -> str: ...
-    def read(self, n: int = ...) -> str: ...
+    def read(self, size: int = ...) -> str: ...
+    def __enter__(self) -> '_RawIOBase': ...
 
-class FileIO(_RawIOBase):
+class FileIO(_RawIOBase, BytesIO):  # type: ignore  # for __enter__
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: Would this not be necessary if you used self types for these `__enter__` methods?

File: stdlib/2/io.pyi [hunk 5]
```diff
@@ -7,98 +7,28 @@
 from typing import List, BinaryIO, TextIO, IO, overload, Iterator, Iterable, Any, Union, Optional
 import _io
 
-DEFAULT_BUFFER_SIZE = 0
+from _io import (DEFAULT_BUFFER_SIZE, BlockingIOError, UnsupportedOperation,
```
Comment (JelleZijlstra) [inline stdlib/2/io.pyi]: These need `x as x` imports to make clear that they're re-exported.

File: stdlib/2/_io.pyi [hunk 6]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
-
-def open(file: Union[int, str], mode: str = ...) -> _IOBase: ...
+    def __init__(self, buffer: IO, encoding: unicode = ...,
+                 errors: unicode = ..., newline: unicode = ...,
+                 line_buffering: bool = ...,
+                 write_through: bool = ...) -> None: ...
+    def __enter__(self) -> 'TextIOWrapper': ...
+
+def open(file: Union[str, unicode, int],
+         mode: unicode = ..., buffering: int = ..., encoding: unicode = ...,
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: encoding, errors, and newline are Optional according to the docs.

File: stdlib/2/_io.pyi [hunk 7]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
+    def writelines(self, lines: Iterable[bytes]) -> None: ...
+    def read1(self, size: int) -> bytes: ...
+    def __iter__(self) -> Iterator[bytes]: ...
+    def next(self) -> bytes: ...
+    def __enter__(self) -> 'BytesIO': ...
 
 class _RawIOBase(_IOBase):
     def readall(self) -> str: ...
-    def read(self, n: int = ...) -> str: ...
+    def read(self, size: int = ...) -> str: ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: The argument is actually called "n" both in the docs and in the C code.

File: stdlib/2/_io.pyi [hunk 8]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
-    def __iter__(self) -> Iterator[bytes]: ...
-    def __enter__(self) -> '_IOBase': ...
-    def __exit__(self, t: Optional[Type[BaseException]], value: Optional[BaseException],
-                 # TODO: traceback should be TracebackType but that's defined in types
-                 traceback: Optional[Any]) -> bool: ...
 
 class _BufferedIOBase(_IOBase):
-    def read1(self, n: int) -> str: ...
-    def read(self, n: int = ...) -> str: ...
+    def read1(self, size: int) -> bytes: ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: These should also be `n` according to the docs.

File: stdlib/2/_io.pyi [hunk 9]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
+    def read(self, size: int = ...) -> bytes: ...
     def readinto(self, buffer: bytearray) -> int: ...
-    def write(self, s: str) -> int: ...
+    def write(self, s: bytes) -> int: ...
     def detach(self) -> "_BufferedIOBase": ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: This is supposed to return the underlying raw stream, so I don't think the return type is right.

File: stdlib/2/_io.pyi [hunk 10]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
     def readall(self) -> str: ...
-    def read(self, n: int = ...) -> str: ...
+    def read(self, size: int = ...) -> str: ...
+    def __enter__(self) -> '_RawIOBase': ...
 
-class FileIO(_RawIOBase):
+class FileIO(_RawIOBase, BytesIO):  # type: ignore  # for __enter__
     mode = ...  # type: str
     closefd = ...  # type: bool
+    def __init__(self, file: str, mode: str = ...) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: The constructor argument is `name` according to the docs, not `file`.

File: stdlib/2/_io.pyi [hunk 11]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
     def readinto(self, buffer: bytearray)-> int: ...
     def write(self, pbuf: str) -> int: ...
+    def __enter__(self) -> 'FileIO': ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: This is just `_IOBase`'s `__enter__`, so I don't think it needs to be here.

File: stdlib/2/_io.pyi [hunk 12]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
+    def read(self, size: int = ...) -> bytes: ...
     def readinto(self, buffer: bytearray) -> int: ...
-    def write(self, s: str) -> int: ...
+    def write(self, s: bytes) -> int: ...
     def detach(self) -> "_BufferedIOBase": ...
+    def __enter__(self) -> '_BufferedIOBase': ...
 
 class BufferedRWPair(_BufferedIOBase):
-    def peek(self, n: int = ...) -> str: ...
+    def __init__(self, reader: _RawIOBase, writer: _RawIOBase) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: This also takes `buffer_size` and `max_buffer_size` arguments.

File: stdlib/2/_io.pyi [hunk 13]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
     def getstate(self) -> Tuple[Any, int]: ...
     def setstate(self, state: Tuple[Any, int]) -> None: ...
     def reset(self) -> None: ...
 
-class _TextIOBase(_IOBase):
+
+# Note: In the actual _io.py, _TextIOBase inherits from _IOBase.
+class _TextIOBase(TextIO):
     errors = ...  # type: Optional[str]
     newlines = ...  # type: Union[str, unicode]
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: According to the docs it can also be None or a tuple of strings.

File: stdlib/2/_io.pyi [hunk 14]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
-        raise UnsupportedOperation
+    encoding = ...  # type: str
+    @property
+    def closed(self) -> bool: ...
+    def _checkClosed(self) -> None: ...
+    def _checkReadable(self) -> None: ...
+    def _checkSeekable(self) -> None: ...
+    def _checkWritable(self) -> None: ...
+    def close(self) -> None: ...
+    def detach(self) -> None: ...
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: This returns the underlying BufferedIOBase, not None.

File: stdlib/2/_io.pyi [hunk 15]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
 
 class TextIOWrapper(_TextIOBase):
     name = ...  # type: str
     line_buffering = ...  # type: bool
-    buffer = ...  # type: str
+    buffer = ...  # type: BinaryIO
     _CHUNK_SIZE = ...  # type: int
-
-def open(file: Union[int, str], mode: str = ...) -> _IOBase: ...
+    def __init__(self, buffer: IO, encoding: unicode = ...,
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: I think this should be `IO[bytes]`.

File: stdlib/2/_io.pyi [hunk 16]
```diff
@@ -19,91 +19,149 @@ class _IOBase(BinaryIO):
...
+                 errors: unicode = ..., newline: unicode = ...,
```
Comment (JelleZijlstra) [inline stdlib/2/_io.pyi]: Further up you had encoding and errors attributes as `str` on `_TextIOBase`.

File: stdlib/2/_io.pyi (hunk 8 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done.

File: stdlib/2/_io.pyi (hunk 1 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done.

File: stdlib/2/_io.pyi (hunk 2 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: My bad. The types got under the wheel when I shifted `__exit__` back and fro.

That said, I'm not sure what we get out of having types on `__exit__`. This function is invoked by the type-checker, not user code (nobody ever does a manual `x.__exit__(a, b,c)`), and the parameters are not originating from user code, either.

Either way, I added types.

File: stdlib/2/_io.pyi (hunk 12 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done.

File: stdlib/2/_io.pyi (hunk 7 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done. The C code seems inconsistent with regards to `n` and `size`. I shouldn't have made these changes in bulk.

(The parameter name probably doesn't matter, since these methods don't allow keyword args. But even so.)

File: stdlib/2/_io.pyi (hunk 4 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done.

File: stdlib/2/_io.pyi (hunk 10 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Looks like the docs are wrong. https://github.com/python/cpython/blob/master/Modules/_io/fileio.c#L199

File: stdlib/2/_io.pyi (hunk 11 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done.

File: stdlib/2/_io.pyi (hunk 13 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Actually, on `_TextIOBase`, this always returns None:

https://github.com/python/cpython/blob/2.7/Modules/_io/textio.c#L104

overloaded below.

File: stdlib/2/_io.pyi (hunk 14 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Had to make this `IO`, since _TextIOBase doesn't inherit from `BufferedIOBase`, and the underlying stream might be either.
`buffer` is `IO` below, too.

File: stdlib/2/_io.pyi (hunk 15 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: No, this works:

`_io.TextIOWrapper(_io.StringIO(u"hello world")).read()`

File: stdlib/2/_io.pyi (hunk 16 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done. These parameters are actually more permissive. They allow `None`, as well.

File: stdlib/2/_io.pyi (hunk 6 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: Done.

File: stdlib/2/io.pyi (hunk 5 above)
Comment (matthiaskramm) [inline stdlib/2/io.pyi]: Done.

**Comment by matthiaskramm** [comment]
Thanks for the thorough review!

File: stdlib/2/_io.pyi (hunk 10 above)
Comment (matthiaskramm) [inline stdlib/2/_io.pyi]: (That was the Python 3 version, actually. Python 2 is the same: https://github.com/python/cpython/blob/2.7/Modules/_io/fileio.c#L183)

**Comment by matthiaskramm** [comment]
Is there a way to tell github that (you think) you addressed all the requested changes on a PR? It seems we occasionally have dangling PRs because it's not clear whether the author is still working on it or not.

**Comment by gvanrossum** [comment]
I don't think so, the convention we follow is for the contributor to send a
comment to the PR saying they're done addressing the feedback.

**Comment by JelleZijlstra** [comment]
That's also something I've missed in GitHub's workflow (similar to "request review" in Phabricator).

I'll take another look at this PR in a few days.

**Comment by gvanrossum** [comment]
Well, "request review" exists -- it's just that "request re-review" doesn't. :-)

**Comment by JelleZijlstra** [comment]
I guess I was stretching the definition of "a few days" :)

Thanks!
